cmd/maestro/server/grpc_broker.go

```diff
@@ -442,6 +473,34 @@ func (bkr *GRPCBroker) OnStatusUpdate(ctx context.Context, eventID, resourceID s
 	)
 }
 
+// PredicateEvent checks if the event should be processed by the current instance
+// by verifying the resource consumer name is in the subscriber list, ensuring the
+// event will be only processed when the consumer is subscribed to the current broker.
+func (bkr *GRPCBroker) PredicateEvent(ctx context.Context, eventID string) (bool, error) {
+	evt, err := bkr.eventService.Get(ctx, eventID)
+	if err != nil {
+		return false, fmt.Errorf("failed to get event %s: %s", eventID, err.Error())
+	}
+
+	// fast return if the event is already reconciled
+	if evt.ReconciledDate != nil {
+		return false, nil
+	}
+
+	resource, svcErr := bkr.resourceService.Get(ctx, evt.SourceID)
+	if svcErr != nil {
+		// if the resource is not found, it indicates the resource has been handled by other instances.
+		if svcErr.Is404() {
+			klog.V(10).Infof("The resource %s has been deleted, mark the event as reconciled", evt.SourceID)
+			return false, nil
```

Comment:
who will mark the event as reconciled? it should be marked by the instance which handle this resource.

Reply:
yes, the event is marked by the instance who handle this event.
This was added to handle this case:
1. resource created and then deleted, but no agent running.
2. agent start, resync the resource.
3. the resource is deleted, but the event will not marked as reconciled.
4. if handle the event again. will get 404 when get resource from db.

Reply:
@clyang82 added logic to mark event as reconciled here.